ext/yajl/yajl_ext.c

```diff
@@ -560,6 +566,385 @@ static VALUE rb_yajl_parser_set_complete_cb(VALUE self, VALUE callback) {
+            } else {
+                rb_raise(cStandardError, "unknown boolean token %s", event.buf);
+            }
+        case yajl_tok_integer:;
+        case yajl_tok_double:;
+            char *buf = (char *)malloc(event.len + 1);
+            buf[event.len] = 0;
+            memcpy(buf, event.buf, event.len);
+
+            VALUE val;
+            if (memchr(buf, '.', event.len) ||
+                memchr(buf, 'e', event.len) ||
+                memchr(buf, 'E', event.len)) {
+                val = rb_float_new(strtod(buf, NULL));
+            } else {
+                val = rb_cstr2inum(buf, 10);
+            }
+            free(buf);
+            
+            return val;
+
+        case yajl_tok_string:;
+        case yajl_tok_string_with_escapes:;
+            return rb_yajl_projector_build_string(parser, event);
+
+        case yajl_tok_eof:;
+            rb_raise(cParseError, "unexpected eof while constructing value");
+
+        default:;
+            assert(0);
+    }
+}
+
+static VALUE rb_yajl_projector_build_string(yajl_event_stream_t parser, yajl_event_t event) {
+    switch (event.token) {
+        case yajl_tok_string:; {
+            VALUE str = rb_str_new(event.buf, event.len);
+            rb_enc_associate(str, utf8Encoding);
+
+            rb_encoding *default_internal_enc = rb_default_internal_encoding();
+            if (default_internal_enc) {
+                str = rb_str_export_to_enc(str, default_internal_enc);
+            }
+
+            return str;
+        }
+
+        case yajl_tok_string_with_escapes:; {
+            //printf("decoding string with escapes\n");
+
+            yajl_buf strBuf = yajl_buf_alloc(parser->funcs);
+            yajl_string_decode(strBuf, (const unsigned char *)event.buf, event.len);
+
+            VALUE str = rb_str_new((const char *)yajl_buf_data(strBuf), yajl_buf_len(strBuf));
+            rb_enc_associate(str, utf8Encoding);
+
+            yajl_buf_free(strBuf);
+
+            rb_encoding *default_internal_enc = rb_default_internal_encoding();
+            if (default_internal_enc) {
+                str = rb_str_export_to_enc(str, default_internal_enc);
+            }
+
+            return str;
+        }
+
+        default:; {
+            assert(0);
+        }
+    }
+}
+
+/*
+ * Document-method: project
+ */
+static VALUE rb_yajl_projector_project(VALUE self, VALUE schema) {
+    yajl_alloc_funcs allocFuncs;
+    yajl_set_default_alloc_funcs(&allocFuncs);
+
+    VALUE stream = rb_iv_get(self, "@stream");
+
+    long buffer_size = FIX2LONG(rb_iv_get(self, "@buffer_size"));
+    VALUE buffer = rb_str_new(0, buffer_size);
+
+    struct yajl_event_stream_s parser = {
+        .funcs = &allocFuncs,
+
+        .stream = stream,
+
+        .buffer = buffer,
+        .offset = buffer_size,
+
+        .lexer = yajl_lex_alloc(&allocFuncs, 0, 1),
+    };
+
+    VALUE result = rb_yajl_projector_filter_subtree(&parser, schema, yajl_event_stream_next(&parser, 1));
```

Comment:
`rb_yajl_projector_filter_subtree` can raise an exception.  In which case the lexer won't get freed.  We should move this around a little to fix the leak.